corehq/ex-submodules/casexml/apps/phone/data_providers/case/livequery.py

```diff
@@ -0,0 +1,208 @@
+        while next_ids:
+            level += 1
+            with timing_context("get_related_indices(level %s)" % level):
+                related = accessor.get_related_indices(list(next_ids), all_ids)
+            if not related:
+                break
+            prev_ids = next_ids
+            next_ids = set()
+            for index in related:
+                indices[index.case_id].append(index)
+                sub_id = index.case_id
+                ref_id = index.referenced_id  # aka parent/host/super
+                relation = index.relationship
+                debug("%s --%s--> %s", sub_id, relation, ref_id)
+                if sub_id in prev_ids:
+                    # traverse to parent/host
+                    if ref_id not in all_ids:  # exclude circular
+                        next_ids.add(ref_id)
+
+                    if (sub_id in live_ids
+                            or (relation == EXTENSION and ref_id in live_ids)
+                            or (relation != EXTENSION and sub_id in open_ids)):
+                        # one of the following is true
+                        # - ref has a live child
+                        # - ref has a live extension
+                        # - sub is the extension of a live case
+                        # - sub is open and is not an extension case
+                        enliven(sub_id)
+                        enliven(ref_id)
+                    else:
+                        # need to know if parent is live before -> live
+                        extensions_by_host[ref_id].add(sub_id)
+                        hosts_by_extension[sub_id].add(ref_id)
+                else:
+                    # traverse to open extension
+                    assert ref_id in prev_ids, (index, prev_ids)
+                    assert relation == EXTENSION, index
+                    if sub_id in all_ids:  # exclude circular
+                        next_ids.add(sub_id)
+
+                    if ref_id in live_ids:
+                        # sub is the extension of a live case
+                        enliven(sub_id)
+                    else:
+                        # need to know if parent is live before -> live
+                        extensions_by_host[ref_id].add(sub_id)
+                        hosts_by_extension[sub_id].add(ref_id)
+
+            debug('next: %r all: %r', next_ids, all_ids)
+            all_ids.update(next_ids)
+
+        # owned, open, not an extension -> live
+        for case_id in open_ids:
+            if case_id not in hosts_by_extension:
+                enliven(case_id)
+
+        # open root nodes and their extensions -> live
+        # TODO store referenced_id open/closed status in case index to avoid this query
+        root_ids = [r for r in extensions_by_host if r not in hosts_by_extension]
+        debug('roots: %r', root_ids)
+        for case_id in accessor.filter_open_case_ids(root_ids):
+            enliven(case_id)
+
+        debug('live: %r', live_ids)
```

Comment:
- Is it necessary to save updated sync logs somewhere around here?
- Possibly related: should this tell the phone to remove no-longer-relevant cases?